libfoundation/src/foundation-java-private.cpp

```diff
@@ -0,0 +1,1353 @@
+{
+    MCAutoStringRef t_class_path;
+    if (!MCJavaClassNameToPathString(p_class_name, &t_class_path))
+        return nullptr;
+    
+    MCAutoStringRefAsCString t_class_cstring, t_method_cstring, t_return_cstring;
+    t_class_cstring . Lock(*t_class_path);
+    t_method_cstring . Lock(p_method_name);
+    t_return_cstring . Lock(p_return);
+    
+    MCJavaDoAttachCurrentThread();
+    
+    jclass t_java_class = s_env->FindClass(*t_class_cstring);
+    
+    void *t_id = nullptr;
+    if (t_java_class != nullptr)
+    {
+        switch (p_call_type)
+        {
+            case MCJavaCallTypeInstance:
+            case MCJavaCallTypeNonVirtual:
+            case MCJavaCallTypeStatic:
+            {
+                MCAutoStringRef t_sig;
+                if (!MCStringFormat(&t_sig, "%@%@", p_arguments, p_return))
+                    return nullptr;
+                
+                MCAutoStringRefAsCString t_signature_cstring;
+                if (!t_signature_cstring . Lock(*t_sig))
+                    return nullptr;
+                
+                if (p_call_type == MCJavaCallTypeStatic)
+                {
+                    t_id = s_env->GetStaticMethodID(t_java_class, *t_method_cstring, *t_signature_cstring);
+                    break;
+                }
+                
+                t_id = s_env->GetMethodID(t_java_class, *t_method_cstring, *t_signature_cstring);
+                break;
+            }
+            case MCJavaCallTypeConstructor:
+            {
+                // Constructors are called with a void return type, and using
+                // the method name <init>
+                MCAutoStringRef t_sig;
+                if (!MCStringFormat(&t_sig, "%@%s", p_arguments, "V"))
+                    return nullptr;
+            
+                MCAutoStringRefAsCString t_signature_cstring;
+                if (!t_signature_cstring . Lock(*t_sig))
+                    return nullptr;
+    
+                t_id = s_env->GetMethodID(t_java_class, "<init>", *t_signature_cstring);
+                break;
+            }
+            case MCJavaCallTypeGetter:
+                t_id = s_env -> GetFieldID(t_java_class, *t_method_cstring, *t_return_cstring);
+                break;
+            case MCJavaCallTypeStaticGetter:
+                t_id = s_env -> GetStaticFieldID(t_java_class, *t_method_cstring, *t_return_cstring);
+                break;
+                
+            case MCJavaCallTypeStaticSetter:
+            case MCJavaCallTypeSetter:
+            {
+                // Remove brackets from arg string to find field type
+                MCAutoStringRef t_args;
+                if (!MCStringCopySubstring(p_arguments,
+                                           MCRangeMake(1, MCStringGetLength(p_arguments) - 2),
+                                           &t_args))
+                    return nullptr;
```

Comment:
This "get the arg string without its parentheses" operation is repeated in a couple of places, so consider splitting it out into a helper function.